File: WoofAndroid/presentation/grid/src/main/kotlin/com/perrystreet/woof/presentation/grid/viewmodel/GridViewModel.kt
package com.perrystreet.woof.presentation.grid.viewmodel

import com.perrystreet.woof.models.dog.DogsFeed
import com.perrystreet.woof.presentation.common.viewmodel.StateProducingViewModel
import com.perrystreet.woof.presentation.grid.mapper.DogDomainToCellUIModelMapper
import com.perrystreet.woof.presentation.grid.uimodel.DogCellUIModel
import com.perrystreet.woof.usecase.dogs.GetDogsFeedUseCase
import com.perrystreet.woof.usecase.dogs.LoadNextDogsPageUseCase
import com.perrystreet.woof.utils.guard
import io.reactivex.rxjava3.kotlin.plusAssign
import org.koin.android.annotation.KoinViewModel
import java.util.Optional

@KoinViewModel
class GridViewModel(
    private val getDogsFeedUseCase: GetDogsFeedUseCase,
    private val loadNextDogsPageUseCase: LoadNextDogsPageUseCase,
    private val cellMapper: DogDomainToCellUIModelMapper,
) : StateProducingViewModel<GridViewModel.State>(initialValue = State.Loading) {

    sealed class State {
        data object Loading : State()

        data class Loaded(
            val cells: List<DogCellUIModel>,
            val isLoadingMore: Boolean,
            val hasMore: Boolean,
        ) : State() {
            val loadingMoreCellCount: Int
                get() = when (isLoadingMore) {
                    true -> LoadingMoreCellCount
                    false -> 0
                }
        }

        data object Error : State()
    }

    private var isLoadingPage = false

    override fun onFirstAppear() {
        disposables += getDogsFeedUseCase().subscribe { feed -> onFeedUpdate(feed) }
        loadNextPage()
    }

    fun onCellAppear(cell: DogCellUIModel) {
        val loaded = currentState as? State.Loaded
        guard(loaded != null && loaded.hasMore && !isLoadingPage) { return }
        guard(cell.index >= loaded.cells.size - LoadMoreThreshold) { return }
        loadNextPage()
    }

    fun onRetryTap() {
        _state.onNext(State.Loading)
        loadNextPage()
    }

    private fun onFeedUpdate(feed: DogsFeed) {
        guard(feed.dogs.isNotEmpty()) { return }
        _state.onNext(
            State.Loaded(
                cells = feed.dogs.mapIndexed { index, dog -> cellMapper(dog, index) },
                isLoadingMore = isLoadingPage,
                hasMore = feed.hasMore,
            ),
        )
    }

    private fun loadNextPage() {
        isLoadingPage = true
        updateLoadingMore()
        disposables += loadNextDogsPageUseCase().subscribe(
            {
                isLoadingPage = false
                updateLoadingMore()
            },
            { error ->
                isLoadingPage = false
                onPageError(error)
            },
        )
    }

    private fun updateLoadingMore() {
        val loaded = currentState as? State.Loaded
        guard(loaded != null) { return }
        _state.onNext(loaded.copy(isLoadingMore = isLoadingPage))
    }

    private fun onPageError(error: Throwable) {
        when (currentState) {
            is State.Loaded -> {
                updateLoadingMore()
                mutableError.onNext(Optional.of(error))
            }
            State.Loading, State.Error -> _state.onNext(State.Error)
        }
    }

    private companion object {
        const val LoadMoreThreshold = 12
        const val LoadingMoreCellCount = 6
    }
}

File: WoofIOS/SwiftPackages/Presentation/Grid/Sources/PresentationGrid/ViewModel/GridViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase
import Utils

@Factory
public final class GridViewModel: StateProducingViewModel<GridViewModel.State, DogsError> {
    private let getDogsFeedUseCase: GetDogsFeedUseCase
    private let loadNextDogsPageUseCase: LoadNextDogsPageUseCase
    private let cellMapper: DogDomainToCellUIModelMapper

    private var isLoadingPage = false

    private static let loadMoreThreshold = 12
    static let loadingMoreCellCount = 6

    init(
        getDogsFeedUseCase: GetDogsFeedUseCase,
        loadNextDogsPageUseCase: LoadNextDogsPageUseCase,
        cellMapper: DogDomainToCellUIModelMapper
    ) {
        self.getDogsFeedUseCase = getDogsFeedUseCase
        self.loadNextDogsPageUseCase = loadNextDogsPageUseCase
        self.cellMapper = cellMapper
        super.init(initialValue: .loading)
    }

    public override func onFirstAppear() {
        getDogsFeedUseCase()
            .sink { [weak self] feed in self?.onFeedUpdate(feed) }
            .store(in: &cancellables)
        loadNextPage()
    }

    public func onCellAppear(_ cell: DogCellUIModel) {
        guard case .loaded(let loaded) = currentState, loaded.hasMore, !isLoadingPage else { return }
        guard cell.index >= loaded.cells.count - Self.loadMoreThreshold else { return }
        loadNextPage()
    }

    public func onRetryTap() {
        setState(.loading)
        loadNextPage()
    }

    private func onFeedUpdate(_ feed: DogsFeed) {
        guard !feed.dogs.isEmpty else { return }
        setState(
            .loaded(
                State.Loaded(
                    cells: feed.dogs.enumerated().map { index, dog in cellMapper(dog: dog, index: index) },
                    isLoadingMore: isLoadingPage,
                    hasMore: feed.hasMore
                )
            )
        )
    }

    private func loadNextPage() {
        isLoadingPage = true
        updateLoadingMore()
        loadNextDogsPageUseCase()
            .pss_sink { [weak self] result in
                guard let self else { return }
                switch result {
                case .success:
                    isLoadingPage = false
                    updateLoadingMore()
                case .failure(let error):
                    isLoadingPage = false
                    onPageError(error)
                }
            }
            .store(in: &cancellables)
    }

    private func updateLoadingMore() {
        guard case .loaded(let loaded) = currentState else { return }
        setState(.loaded(loaded.copy(isLoadingMore: isLoadingPage)))
    }

    private func onPageError(_ error: DogsError) {
        switch currentState {
        case .loaded:
            updateLoadingMore()
            emitError(error)
        case .loading, .error:
            setState(.error)
        }
    }
}

extension GridViewModel {
    public enum State: Equatable {
        case loading
        case loaded(Loaded)
        case error

        public struct Loaded: Equatable {
            public let cells: [DogCellUIModel]
            public let isLoadingMore: Bool
            public let hasMore: Bool

            public var loadingMoreCellCount: Int {
                switch isLoadingMore {
                case true: GridViewModel.loadingMoreCellCount
                case false: 0
                }
            }

            public init(cells: [DogCellUIModel], isLoadingMore: Bool, hasMore: Bool) {
                self.cells = cells
                self.isLoadingMore = isLoadingMore
                self.hasMore = hasMore
            }

            func copy(isLoadingMore: Bool) -> Loaded {
                Loaded(cells: cells, isLoadingMore: isLoadingMore, hasMore: hasMore)
            }
        }
    }
}
